packages/web/src/pages/oauth-login-page/OAuthLoginPage.tsx

```diff
@@ -98,111 +212,259 @@ export const OAuthLoginPage = () => {
     try {
       signedData = await AudiusBackend.signDiscoveryNodeRequest(message)
     } catch {
-      setSubmitError('Something went wrong.')
       return
     }
     const signature = signedData.signature
-
     return `${header}.${payload}.${base64url.encode(signature)}`
   }
 
-  const onFormSubmit = async (e: FormEvent) => {
+  const authAndRedirect = async (account: User) => {
+    const jwt = await formOAuthResponse(account)
+    if (jwt == null) {
+      setIsSubmitting(false)
+      setGeneralSubmitError(messages.miscError)
+      return
+    }
+    const statePart = state != null ? `state=${state}&` : ''
+    const fragment = `#${statePart}token=${jwt}`
+    if (isRedirectValid === true) {
+      window.location.href = `${redirect_uri}${fragment}`
+    }
+  }
+
+  const handleSignInFormSubmit = async (e: FormEvent) => {
     e.preventDefault()
+    clearErrors()
+    setIsSubmitting(true)
     let signInResponse: any
     try {
-      signInResponse = await AudiusBackend.signIn(email, password)
+      signInResponse = await AudiusBackend.signIn(emailInput, passwordInput)
     } catch (err) {
-      setSubmitError('Unknown error')
+      setIsSubmitting(false)
```

Comment:
are there any cases where we would potentially override `isSubmitting` and `generalSubmitError` with two in-flight async calls? reason i ask, is that we could maybe use `useAsync` and `useAsyncFn` to define these callbacks/effects which would isolate the loading and error logic. down to chat about those options.